pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,233 @@ func (q *QueryGVR) Name() string {
 }
 
 // WithVersions checks if an API group with all the specified versions exist.
+// This method can be omitted to query any version.
 func (q *QueryGVR) WithVersions(versions ...string) *QueryGVR {
 	q.versions = versions
 	return q
 }
 
 // WithResource checks if an API group with the specified resource exists.
-// WithVersions needs to be used before calling this function.
+// This method can be omitted to query any resource.
 func (q *QueryGVR) WithResource(resource string) *QueryGVR {
-	q.resource = resource
+	q.resource.set(resource)
 	return q
 }
 
-// Run discovery
+// Run discovery.
 func (q *QueryGVR) Run(config *clusterQueryClientConfig) (bool, error) {
-	if q.group != "" && q.resource != "" && len(q.versions) == 0 {
-		return false, fmt.Errorf("cannot check for group and resource existence without version info; use WithVersion method")
+	if err := q.validate(config); err != nil {
+		return false, fmt.Errorf("GroupVersionResource API query validation: %w", err)
+	}
+
+	var unmatched []string
+
+	switch {
+	case q.versions == nil:
+		// q.WithVersions method was omitted and q.versions has not been set.
+		// Any version that matches group and/or resource is considered a match.
+		unmatchedGroupResource, err := q.unmatchedGroupResource(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupResource: %w", err)
+		}
+		if unmatchedGroupResource != "" {
+			unmatched = append(unmatched, unmatchedGroupResource)
+		}
+	case q.resource.String == "":
+		unmatchedGroupVersions, err := q.unmatchedGroupVersions(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersions: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGroupVersions...)
+	case q.resource.String != "":
+		unmatchedGVRs, err := q.unmatchedGroupVersionResources(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersionResources: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGVRs...)
+	}
+
+	q.unmatchedGVRs = unmatched
+	return len(unmatched) == 0, nil
+}
+
+func (q *QueryGVR) validate(cfg *clusterQueryClientConfig) error {
+	if cfg == nil {
+		return fmt.Errorf("clusterQueryClientConfig must not be nil")
 	}
 
-	var gvrs []schema.GroupVersionResource
-	if len(q.versions) > 0 {
-		for _, v := range q.versions {
-			gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group, Version: v, Resource: q.resource})
+	if q.resource.IsSet && q.resource.String == "" {
+		return ErrInvalidWithResourceMethodArgument
+	}
+
+	if q.versions != nil && containsEmpty(q.versions) {
```

Comment:
should you be validating the versions first and then the resource later?
this `Group("test", "").WithVersions("").WithResource("")` returns `ErrInvalidWithResourceMethodArgument`, it should return `ErrInvalidWithVersionsMethodArgument` if we are immediately returning the error

This also makes me think, should we append the errors, because in the above case we have invalid versions and resource?

Reply:
> should you be validating the versions first and then the resource later?

I suppose given the example you gave, validating versions should come first. However, if the developer used a different order, i.e. `Group("test", "").WithResource("").WithVersions("")`, then I think the resource error should come first. 

But, I don't think it's worth the effort to ensure a "correct" order of error messages.

> should we append the errors, because in the above case we have invalid versions and resource?

I don't mind doing that; I went ahead and updated the error wrapping to include both errors if they both occur.